Added tags for 6 major ocean basins to ocean regions

xylar opened the pull request:
Nearly all ocean regions should now belong to one of: Pacific, Atlantic, Arctic or Indian Oceans

Also, this merge adds a script, `tag_features.py`  for adding a tag to a feature collection.

xylar commented:
The following script demonstrates how `tag_features.py` is used and how the tags were added to each ocean region.

```bash
#!/bin/bash

# Atlantic Ocean
rm features.geojson
for feature in North_Atlantic_Ocean Celtic_Sea Bay_of_Biscay Norwegian_Sea Greenland_Sea \ 
Labrador_Sea Davis_Strait Hudson_Bay Baffin_Bay Gulf_of_St-Lawrence Gulf_of_Mexico \ 
Caribbean_Sea Gulf_of_Guinea South_Atlantic_Ocean "Irish_Sea_and_St._George's_Channel" \ 
Inner_Seas_off_the_West_Coast_of_Scotland Hudson_Strait North_Sea Skaggerak Kattegat \ 
Baltic_Sea Gulf_of_Riga Gulf_of_Finland Gulf_of_Bothnia Strait_of_Gibraltar Balearic_Sea \ 
Mediterranean_Sea_-_Eastern_Basin Mediterranean_Sea_-_Western_Basin Ligurian_Sea \ 
Tyrrhenian_Sea Adriatic_Sea Ionian_Sea Aegean_Sea Sea_of_Marmara Black_Sea Sea_of_Azov \ 
English_Channel Alboran_Sea Bay_of_Fundy Rio_de_La_Plata
do
  ./merge_features.py -f ocean/region/$feature/region.geojson
done
./tag_features.py -f features.geojson -t Atlantic_Ocean
./plot_features.py -f features.geojson -o Atlantic.png -m cyl
./split_features.py -f features.geojson

# Pacific Ocean
rm features.geojson
for feature in North_Pacific_Ocean Gulf_of_Alaska Bering_Sea Sea_of_Okhotsk Japan_Sea \ 
Eastern_China_Sea South_China_Sea Gulf_of_Thailand Singapore_Strait Celebes_Sea \
Gulf_of_Tomini Halamahera_Sea Ceram_Sea Gulf_of_Boni Java_Sea Bali_Sea Flores_Sea \ 
Savu_Sea Arafura_Sea Bismarck_Sea Philippine_Sea Sulu_Sea Molukka_Sea Banda_Sea \ 
Timor_Sea Solomon_Sea Coral_Sea Tasman_Sea Bass_Strait Gulf_of_California \ 
The_Coastal_Waters_of_Southeast_Alaska_and_British_Columbia Makassar_Strait \ South_Pacific_Ocean Yellow_Sea
do
  ./merge_features.py -f ocean/region/$feature/region.geojson
done
./tag_features.py -f features.geojson -t Pacific_Ocean
./plot_features.py -f features.geojson -o Pacific.png -m cyl
./split_features.py -f features.geojson

# Indian Ocean
rm features.geojson
for feature in Indian_Ocean Mozambique_Channel Gulf_of_Suez Gulf_of_Aqaba Red_Sea \ 
Gulf_of_Aden Persian_Gulf Gulf_of_Oman Arabian_Sea Laccadive_Sea Bay_of_Bengal \
Andaman_or_Burma_Sea Malacca_Strait Great_Australian_Bight
do
  ./merge_features.py -f ocean/region/$feature/region.geojson
done
./tag_features.py -f features.geojson -t Indian_Ocean
./plot_features.py -f features.geojson -o Indian.png -m cyl
./split_features.py -f features.geojson

# Arctic Ocean
rm features.geojson
for feature in Arctic_Ocean Barentsz_Sea White_Sea Kara_Sea Laptev_Sea East_Siberian_Sea \ 
Chukchi_Sea Beaufort_Sea Northwestern_Passages
do
  ./merge_features.py -f ocean/region/$feature/region.geojson
done
./tag_features.py -f features.geojson -t Arctic_Ocean
./plot_features.py -f features.geojson -o Arctic.png -m cyl
./split_features.py -f features.geojson
```

xylar commented:
Here are plots of each region as a sanity check:
![arctic](https://cloud.githubusercontent.com/assets/4179064/12965511/04bdf8ae-d057-11e5-9a2f-4d1f55db053b.png)
![atlantic](https://cloud.githubusercontent.com/assets/4179064/12965512/04c31514-d057-11e5-84a8-b950992b767c.png)
![indian](https://cloud.githubusercontent.com/assets/4179064/12965513/04c44dda-d057-11e5-99d2-e6951ff4bf94.png)
![pacific](https://cloud.githubusercontent.com/assets/4179064/12965514/04c49ccc-d057-11e5-9ee2-67588c5e11dc.png)

xylar commented:
@milenaveneziani and @douglasjacobsen, can you both take a look at this and make sure it seems to make sense.  Milena, feel free to merge this when you're comfortable with it and when Doug has signed off.

xylar commented:
@milenaveneziani, to use these tags, you would do something like:
```
$ ./merge_features.py -d ocean -t Atlantic_Ocean
$ mv features.geojson Atlantic.geojson
$ ./plot_features.py -f Atlantic.geojson -t Atlantic.png -m cyl
```
(You might want to edit the name of the feature group in `Atlantic.geojson`.)  You could then use `Atlantic.geojson` to create masks in different ocean configurations using Doug's scripts (which I'm not yet familiar with).

milenaveneziani commented:
@xylar: I had not seen this one! That's a lot of work, thanks so much!
Concerning my previous comment on the other PR (#18) that was closed earlier today, I would still define the Indian and Pacific Oceans as I mentioned there, which means that:
1) for the Indian Ocean I would include the tmp versions of Great_Australian_Bight and Bass_Strait (defined in my comment to PR #18), and *not* include the Persian_Sea, Red_Sea, Gulf_of_Suez and Gulf_Aqaba in it;
2) for the Pacific Ocean I would just exclude Bass_strait;
3) for the Atlantic Ocean I would exclude the Sea_of_Marmara, Black_Sea, and Sea_of_Azov. I would personally also exclude the Mediterranean, but I see that it was included in POP. Perhaps I'll ask for feedback on this one;
4) no changes for the Arctic.

xylar commented:
@milenaveneziani, okay, that's fine.  I can make all these changes for this PR.

I will also make this comment in confluence so we can discuss, but my feeling is that the regions should not overlap to the extent possible so you're not double counting parts of the Tasmin Sea in the Southern Ocean and in the Pacific.  It would be easy enough to exclude the Tasmin Sea and others from the Southern Ocean instead.  Do you prefer to have the overlap?

milenaveneziani commented:
wow, between me working late and you starting early, the 8-hour difference is canceled.. :)
The reason why I prefer the overlap is that this way we do not compromise the (small) region definition. What really matters is how the region is defined, so if say the Tasman Sea is considered to extend that much to the south, I wouldn't chop it just because we changed SO definition. Now, perhaps the Tasman Sea relevant to oceanography is different from what we have in the original version, but it could go either way (could be smaller than what it is now for example).

xylar commented:
@milenaveneziani, how handy that you're still up!  I was expecting to wait 8 hours for a reply.

How do you feel about me chopping a little bite out of the Southern Ocean so the Tasman Sea, Great Australian Bight and Bass Strait are unchanged from the original definition?  As I said, that's not hard to do.

milenaveneziani commented:
I don't know, I think it's quite a bit. I am really not worried about overlaps (it's also not a requirement on the design document), but you should definitely post something on confluence to see what other people think. If you remember from some of our discussions in past meetings, we even mentioned that we could have different definitions of the same region (say have a few SO regions, all grouped in the SO group).

milenaveneziani commented:
oh, I just read this part of your comment more carefully:
> regions should not overlap to the extent possible so you're not double counting parts 
> of the Tasmin Sea in the Southern Ocean and in the Pacific.
no no, this is not what I am advocating for.. If you read my comment on the closed PR (18), you see that I mean to keep say the Tasman Sea as it is, *but* I want to chop it in 2 when it comes to defining the SO and the Pacific. Do you see what I mean?

xylar commented:
@milenaveneziani, Okay, I see now.  I'll take care of that, then.

milenaveneziani commented:
thanks for doing this, @xylar. and do let me know if I can do something. Time for bed now.

xylar commented:
@milenaveneziani, okay, I think I have set up the basins as you wanted.  First, I split the three seas south of Australia into a northern and southern part.  The northern parts are either part of the Indian Basin (for the Great Australian Bight) or the Pacific Basin (for the other two) and the southern parts are part of the Southern Ocean Basin.  I cut out the overlap of the three seas from the Southern Ocean definition.  (Tasmania and part of New Zealand are counted as part of the Southern Ocean right now, but that's likely harmless.)

The tags now end in `Basin` because it is also useful to have tags for the individual regions that would otherwise have the same name (e.g. the feature `Arctic_Ocean` has both tags `Arctic_Ocean` and `Arctic_Basin`).

Updating my example usage of these tags from above:
```bash
./merge_features.py -d ocean -t Atlantic_Basin
mv features.geojson Atlantic_Basin.geojson
./plot_features.py -f Atlantic_Basin.geojson -t Atlantic_Basin.png -m cyl
```

xylar commented:
For full documentation, this is the script I used to split the 3 seas south of Australia and to cut their overlap out of the Southern Ocean:
```bash
#!/bin/bash

cp ocean/region/Southern_Ocean/region.geojson maskSouth.geojson
cp ocean/region/Southern_Ocean/region.geojson maskNorth.geojson
# make a shape that is everything but the Southern Ocean
vim maskNorth.geojson

for feature in Tasman_Sea Bass_Strait Great_Australian_Bight
do
  rm features.geojson
  ./merge_features.py -f ocean/region/$feature/region.geojson
  mv features.geojson ${feature}.geojson
  ./difference_features.py -f ${feature}.geojson -m maskSouth.geojson
  mv features.geojson ${feature}_North.geojson
  ./difference_features.py -f ${feature}.geojson -m maskNorth.geojson
  mv features.geojson ${feature}_South.geojson
  # rename the features manually (easier than a script)
  vim ${feature}_North.geojson ${feature}_South.geojson
  ./split_features.py -f ${feature}_North.geojson
  ./split_features.py -f ${feature}_South.geojson
done

rm features.geojson
./merge_features.py -f ocean/region/Southern_Ocean/region.geojson
mv features.geojson SouthernOcean.geojson
for feature in Tasman_Sea Bass_Strait Great_Australian_Bight
do
  ./difference_features.py -f SouthernOcean.geojson -m ocean/region/$feature/region.geojson
  mv features.geojson SouthernOcean.geojson
done
./plot_features.py -f SouthernOcean.geojson -o SouthernOcean.png -m southpole
./split_features.py -f SouthernOcean.geojson
```

xylar commented:
...And here is the script I used to add the tags to the basins:
```bash
#!/bin/bash

# Mediterranean Sea
rm features.geojson
for feature in Strait_of_Gibraltar Balearic_Sea Mediterranean_Sea_-_Eastern_Basin \
    Mediterranean_Sea_-_Western_Basin Ligurian_Sea Tyrrhenian_Sea Adriatic_Sea \
    Ionian_Sea Aegean_Sea Sea_of_Marmara Black_Sea Sea_of_Azov \
    Alboran_Sea
do
  ./merge_features.py -f ocean/region/$feature/region.geojson
done
./tag_features.py -f features.geojson -t Mediterranean_Basin
./plot_features_tweaked.py -f features.geojson -o Mediterranean.png -m cyl
./split_features.py -f features.geojson

# Atlantic Ocean
rm features.geojson
for feature in North_Atlantic_Ocean Celtic_Sea Bay_of_Biscay Norwegian_Sea \
    Greenland_Sea Labrador_Sea Davis_Strait Hudson_Bay Baffin_Bay \
    Gulf_of_St-Lawrence Gulf_of_Mexico Caribbean_Sea Gulf_of_Guinea South_Atlantic_Ocean \
    Irish_Sea_and_St_Georges_Channel \
    Inner_Seas_off_the_West_Coast_of_Scotland Hudson_Strait \
    North_Sea Skaggerak Kattegat Baltic_Sea Gulf_of_Riga Gulf_of_Finland Gulf_of_Bothnia \
    English_Channel Bay_of_Fundy Rio_de_La_Plata
do
  ./merge_features.py -f ocean/region/$feature/region.geojson
done
./tag_features.py -f features.geojson -t Atlantic_Basin
./plot_features_tweaked.py -f features.geojson -o Atlantic.png -m cyl
./split_features.py -f features.geojson

# Pacific Ocean
rm features.geojson
for feature in North_Pacific_Ocean Gulf_of_Alaska Bering_Sea Sea_of_Okhotsk Japan_Sea \
    Eastern_China_Sea South_China_Sea Gulf_of_Thailand Singapore_Strait Celebes_Sea \
    Gulf_of_Tomini Halamahera_Sea Ceram_Sea Gulf_of_Boni Java_Sea Bali_Sea Flores_Sea \
    Savu_Sea Arafura_Sea Bismarck_Sea Philippine_Sea Sulu_Sea Molukka_Sea Banda_Sea \
    Timor_Sea Solomon_Sea Coral_Sea Tasman_Sea_North Gulf_of_California \
    The_Coastal_Waters_of_Southeast_Alaska_and_British_Columbia Makassar_Strait \
    South_Pacific_Ocean Yellow_Sea
do
  ./merge_features.py -f ocean/region/$feature/region.geojson
done
./tag_features.py -f features.geojson -t Pacific_Basin
./plot_features_tweaked.py -f features.geojson -o Pacific.png -m cyl
./split_features.py -f features.geojson

# Indian Ocean
rm features.geojson
for feature in Indian_Ocean Mozambique_Channel Gulf_of_Suez Gulf_of_Aqaba Red_Sea \
    Gulf_of_Aden Persian_Gulf Gulf_of_Oman Arabian_Sea Laccadive_Sea Bay_of_Bengal \
    Andaman_or_Burma_Sea Malacca_Strait Great_Australian_Bight_North Bass_Strait_North
do
  ./merge_features.py -f ocean/region/$feature/region.geojson
done
./tag_features.py -f features.geojson -t Indian_Basin
./plot_features_tweaked.py -f features.geojson -o Indian.png -m cyl
./split_features.py -f features.geojson

# Arctic Ocean
rm features.geojson
for feature in Arctic_Ocean Barentsz_Sea White_Sea Kara_Sea Laptev_Sea \
    East_Siberian_Sea Chukchi_Sea Beaufort_Sea Northwestern_Passages
do
  ./merge_features.py -f ocean/region/$feature/region.geojson
done
./tag_features.py -f features.geojson -t Arctic_Basin
./plot_features_tweaked.py -f features.geojson -o Arctic.png -m cyl
./split_features.py -f features.geojson

# Southern Ocean
rm features.geojson
for feature in Southern_Ocean Great_Australian_Bight_South Bass_Strait_South Tasman_Sea_South
do
  ./merge_features.py -f ocean/region/$feature/region.geojson
done
./tag_features.py -f features.geojson -t Southern_Ocean_Basin
./plot_features_tweaked.py -f features.geojson -o SouthernOcean.png -m cyl
./split_features.py -f features.geojson
```

xylar commented:
Here are plots of each of the 6 basins, again including poorly place names at *very* high resolution:

![arctic](https://cloud.githubusercontent.com/assets/4179064/12974705/4621945e-d0b6-11e5-94f0-f862a235cca0.png)
![atlantic](https://cloud.githubusercontent.com/assets/4179064/12974708/462492c6-d0b6-11e5-99ba-b68e33d4810e.png)
![indian](https://cloud.githubusercontent.com/assets/4179064/12974710/4624ba30-d0b6-11e5-8997-58b8e8846a14.png)
![mediterranean](https://cloud.githubusercontent.com/assets/4179064/12974707/46247a66-d0b6-11e5-9f18-bae29d491793.png)
![pacific](https://cloud.githubusercontent.com/assets/4179064/12974709/46249b0e-d0b6-11e5-8ad0-523e49acfc73.png)
![southernocean](https://cloud.githubusercontent.com/assets/4179064/12974706/4623f96a-d0b6-11e5-9988-f9e03b315d39.png)

xylar commented:
Oops, I forgot to take the four seas around the Arabian Peninsula out.  The last commit fixes this, and updates the `./tag_features.py` script to also be able to remove tags.

Here is the corrected Indian Basin.

![indian_basin](https://cloud.githubusercontent.com/assets/4179064/12975131/8a349a26-d0b9-11e5-86e0-f82ce2f6f482.png)

xylar commented:
@milenaveneziani, I can take the Black Sea, etc. out of the Mediterranean if you like.  From a practical standpoint, these are removed from the ocean grid anyway, so I don't think this is likely to matter either way.

douglasjacobsen commented:
Just for future PRs, I think we should separate PRs that add / modify scripts from those that add / modify regions. Unless the script PR requires changes to the regions at least. That way different people can review them independently and keep the PRs nice and clean (and you can do things like modify region definitions relative to a PR conversation without having to worry about the script changes you had made).

douglasjacobsen commented on tag_features.py:
```diff
@@ -0,0 +1,73 @@
+parser.add_argument("-t", "--tag", dest="tag", help="Tag to add to all features", 
+                    metavar="TAG", required=True)
+parser.add_argument("-r", "--remove", dest="remove", action='store_true',
+                    help="Use this tag to signal removing a tag instead of adding")
```
Can you change this to "Use this flag to signal.." ? Just to keep argument flags separate from the feature tags you're trying to modify here?

douglasjacobsen commented:
@xylar Thanks for this. The tag_features.py script looks / works good. I think as far as the python portions of this PR it's ready to merge. I'll let @milenaveneziani be the one who decides about the geojson portions of this PR.

douglasjacobsen commented:
Sorry, I should say it's ready to merge with the one minor fix I commented about.

milenaveneziani commented:
Hi @xylar, thanks for doing this. Looks great. A couple of questions:
- does it matter that the Mediterranean seas are tagged as Atlantic ocean, yet your plot of the Atlantic does not have it? Also, Bass Strait was included in the Indian Ocean but tagged as part of the Pacific. Not sure if all this matter.
- yes, I'd remove the Black Sea (and Marmara Strait and Sea of Azov) from the Mediterranean, but as you say it's not a priority right now.

xylar commented:
@milenaveneziani, are you sure you're looking at the right version.  I'll double check but I think I fixed the tags on the Mediterranean seas and the Bass Strait.  I'll take out the Black Sea and its neighbors.

xylar commented:
@douglasjacobsen, I will keep new scripts separate from feature changes in the future.  This PR evolved beyond my initial intent but I could have kept it separate nonetheless.  Sorry about that.

xylar commented on ocean/region/Bass_Strait_North/region.geojson:
```diff
@@ -3,9 +3,9 @@
+			"name": "Bass Strait North",
 			"component": "ocean",
-			"tags": "",
+			"tags": "Bass_Strait;Indian_Basin",
```
@milenaveneziani, the Bass Strait looks okay here.  Maybe you need to do a `git fetch xylar/geometry_features` and then a `git reset --hard xylar/geometry_features/tag_ocean_basins`?

xylar commented:
@douglasjacobsen, I made yours small change.

milenaveneziani commented:
sorry, @xylar. I was looking at the Files changed page and needed to refresh it!

douglasjacobsen commented:
@xylar No problem, that happens a lot. :smile:

xylar commented:
@milenaveneziani, okay, I just removed the 3 seas from the Mediterranean Basin as you requested.  I guess things are ready from my end...

milenaveneziani commented:
ok, thanks. I'll check it again this afternoon (I'm at the CESM workshop) and merge it.
Any idea why cartopy needs to download additional packages for the plot_features script to work? It would be great if I could get it to work on the setup I have on IC machines, but right now it doesn't accept the downloading step.

xylar commented:
@milenaveneziani, oops, I see that I missed the Bristol Channel in the Atlantic.  I will look for others that i missed as well.

xylar commented:
@milenaveneziani, Yes, Cartopy needs to download the coastlines and land polygons (I think I mentioned that in my previous comment about this issue).  You could figure out what it is trying to download and download it manually and put it into the appropriate directory under anaconda if it's bothering you.  Presumably you have the right files downloaded on your laptop if you got the plot script to work there.  It's just a question of tracking down what shapefiles are under the cartopy directory within the anaconda setup you're using on your laptop and putting those in the same place in your anaconda directory on IC.  X forwarding is far too slow for me to try doing this myself from here, so I'm afraid you're on your own.

xylar commented:
@milenaveneziani, alternatively you could comment out the plotting of coastlines and land masses in the plotting script on IC, or maybe even change the resolution of the coastlines to be the coarsest option (which I think is downloaded by default with cartopy).  Then it might work on IC.

xylar commented:
Okay, I found 3 tiny regions that I had missed tagging.  I really think I got them all now...

milenaveneziani commented:
I was able to recreate the basins you showed above, @xylar.
A couple of things:
1) should we save the basins created this way somewhere?
2) the syntax for the merge_features script is slightly different from the example above. For example:
./merge_features.py -d ocean/region/ -t Arctic_Basin  (full path=ocean/region, and need to use 'Basin' in tag's name)

Finally, thanks for the info about cartopy: I was able to set up the git repo on my laptop.

xylar commented:
Thanks @milenaveneziani.
> 1) should we save the basins created this way (and also the MOC_Basins) somewhere?

No, the idea is that we definitely don't store the combined basins because then they have to be updated every time we change individual region definitions.  Rather, we store a driver script for creating them, and generate them when needed from this repository

> 2) the syntax for the merge_features script is slightly different from the example above. For example:
> `./merge_features.py -d ocean -t Arctic_Basin` (need to use 'Basin' in tag's name).

I believe you were using my earliest example.  When I changed the tags to be `*_Basin`, I updated the procedure above.

> More importantly, should we have a script that makes the basins similar to what we have for the MOC basins?

Yes, we absolutely do need to add such a script.  I'll make a separate PR for that shortly.
